aws_lambda_builders/workflows/java_gradle/DESIGN.md

```diff
@@ -0,0 +1,117 @@
+create distribution ZIP's by default. We can further provide the `--uber-jar`
+option for customers to force the creation of an uber JAR if necessary. This
+distribution will only contain the libraries needed at runtime (i.e. the runtime
+classpath libraries).
+
+Finally, a `--use-container` option will be provided to address potential issues
+with environment-dependent build scripts (see Challenges).
+
+## Challenges
+
+Java bytecode can only run on the same or newer version of the JVM for which
+it was compiled for. For example Java 8 bytecode can run a JVM that is at
+least version 8, but bytecode targetting Java 9 cannot run on a Java 8 VM.
+This is further complicated by the fact that a newer JDK can generate code to
+be run on an older VM if configured using the `targetCompatibility` and
+`sourceCompatibility` properties of the Java plugin. Therefore, it is not
+sufficient to check the version of the local JDK, nor is it possible to check
+the value set for `targetCompatibility` or `sourceCompatibility` since it can
+be local to the compile/build task. At best, we can check if the local
+version of the JDK is newer than Java 8 and emit a warning that the built
+artifact may not run in Lambda.
+
+Gradle projects are configured using `build.gradle` build scripts. These are
+executable files authored in either Groovy or since 5.0, Kotlin, and using the
+Gradle DSL. This presents a similar problem to `setup.py` in the Python world in
+that arbitrary logic can be executed during build time that could affect both
+how the customer's artifact is built, and which dependencies are chosen.
+
+To solve the issue where the build environment could affect the `build.gradle`
+script, a `--use-container` option will be provided, to give the option of
+building in an environment that closely mimics the Lambda environment.
+
+## Implementation
+
+We leverage Gradle to do all the heavy lifting for executing the `build.gradle`
+script which will resolve and download the dependencies and build the project.
+To create the distribution ZIP, we insert a new Gradle `task` that packages the
+compiled classes, resouces, and dependencies into a ZIP.
+
+### sam build
+
+#### Step 1: Check Java version and emit warning
+
+Check whether the local JDK version is <= Java 8, and if it is not, emit a
+warning that the built artifact may not run in Lambda unless a) the project is
+properly configured (i.e. using `targetCompatibility`) or b) the project is
+build within the Lambda-compatibile container.
+
+#### Step 2: Synthesize a new build script
+
+There is no standard task in Gradle to create a distribution ZIP (or uber JAR).
+We must synthesize a new build script that we will run to actually produce the
+Lambda artifact. This is so we can insert a new task that will assemble the
+distribution ZIP. The new script is identical to the customer's original
+`build.gradle` except that it imports our own `sam-ext.gradle` file provided by
```

Comment:
Did you get to look at the initScript system to see if we could leave their file alone? Otherwise we have to also worry about .kts files

Reply:
I did. I basically tried to adapt the given `sam-ext.gradle` as an init script but since we declare `dependsOn` on `build` and `runtimeClasspath`, it gives an error since they're not in scope until `build.gradle` is executed. Happy to discuss if you know a way around this.

+1 on finding `.kts` files for this approach.

Reply:
You may just have to use the actual API instead of relying on things in scope. Like `tasks.findByName("build")`, but I admit I have not had a chance to look

Reply:
Not 100% sure this is the right thing todo. Editing the script seems wrong. Can we provide a plugin that customers should use and then expect it to be there? 

I don't know a ton about gradle but will do some more reading to see what other options we may have here.